crates/wash-lib/src/build/provider.rs

```diff
@@ -0,0 +1,132 @@
+/// and operating system using provided configuration.
+pub(crate) async fn build_provider(
+    provider_config: &ProviderConfig,
+    language_config: &LanguageConfig,
+    common_config: &CommonConfig,
+    signing_config: Option<&SignConfig>,
+) -> Result<PathBuf> {
+    let LanguageConfig::Rust(rust_config) = language_config else {
+        bail!("Unsupported language for provider: {:?}", language_config)
+    };
+    let mut command = match rust_config.cargo_path.as_ref() {
+        Some(path) => process::Command::new(path),
+        None => process::Command::new("cargo"),
+    };
+
+    // Change directory into the project directory
+    std::env::set_current_dir(&common_config.path)?;
+
+    // Build for a specified target if provided, or the default rust target
+    let build_args = if let Some(override_target) = &provider_config.target {
+        vec!["build", "--release", "--target", override_target]
+    } else {
+        vec!["build", "--release"]
+    };
+    let result = command.args(build_args).status().map_err(|e| {
+        if e.kind() == ErrorKind::NotFound {
+            anyhow!("{:?} command is not found", command.get_program())
+        } else {
+            anyhow!(e)
+        }
+    })?;
+
+    if !result.success() {
+        bail!("Compiling provider failed: {}", result.to_string())
+    }
+
+    let metadata = cargo_metadata::MetadataCommand::new().no_deps().exec()?;
+    let bin_name = if let Some(bin_name) = &provider_config.bin_name {
+        bin_name.to_string()
+    } else {
+        // Discover the binary name from the metadata
+        metadata
+            .packages
+            .iter()
+            .find_map(|p| {
+                p.targets.iter().find_map(|t| {
+                    if t.kind.iter().any(|k| k == "bin") {
+                        Some(t.name.clone())
+                    } else {
+                        None
+                    }
+                })
+            }).context("Could not infer provider binary name in metadata, please specify under provider.bin_name")?
+    };
+    let mut provider_path_buf = rust_config
+        .target_path
+        .clone()
+        .unwrap_or_else(|| PathBuf::from(metadata.target_directory.as_std_path()));
+    if let Some(override_target) = &provider_config.target {
+        provider_path_buf.push(override_target);
+    }
+    provider_path_buf.push("release");
+    provider_path_buf.push(&bin_name);
+
+    let provider_bytes = tokio::fs::read(&provider_path_buf).await?;
+
+    let mut par = create_provider_archive(
+        ParCreateArgs {
+            capid: provider_config.package.clone().unwrap_or_default(),
+            vendor: provider_config.vendor.to_string(),
+            revision: Some(common_config.revision),
+            version: Some(common_config.version.to_string()),
+            schema: None,
+            name: common_config.name.to_string(),
+            arch: detect_arch(),
+        },
+        &provider_bytes,
+    )?;
```

Comment:
`.context` would be nice here -- IIRC `create_provider_archive` doesn't have very descriptive errors